Adds end-of-road stop conditions to MaliputRailcar.

liangfok opened the pull request:
This PR is based on https://github.com/RobotLocomotion/drake/pull/5552#issuecomment-288531943. `MaliputRailcar`'s speed is only zero'ed by `CalcUnrestrictedUpdate()`.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/5600)
<!-- Reviewable:end -->

liangfok commented:
+@jadecastro for feature review please.
+@jwnimmer-tri for car review please.
Thanks.

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5600#-:-KftI1O33782YwdpedS5:b-d36nmx)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Finished a pass. PTAL.

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/automotive/maliput_railcar.cc, line 371 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5600#-KfwytBvGE8oX_1LqI42:-KfwytBvGE8oX_1LqI43:b-ypynst) ([raw file](https://github.com/robotlocomotion/drake/blob/a536a916b8c7db975c1b4c777655f475f579f9af/drake/automotive/maliput_railcar.cc#L371)):*
> ```cpp
>   DRAKE_ASSERT(next_railcar_state != nullptr);
> 
>   // Sets the speed to be zero if the car is at the end of the road.
> ```

BTW, `at` --> `after`

---

*[drake/automotive/maliput_railcar.cc, line 374 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5600#-KfwxuVbkIkPJlHdEpAP:-KfwxuVcsiGW7y3nF2JT:bhanq8f) ([raw file](https://github.com/robotlocomotion/drake/blob/a536a916b8c7db975c1b4c777655f475f579f9af/drake/automotive/maliput_railcar.cc#L374)):*
> ```cpp
>   if (current_lane_direction.with_s) {
>     const int num_branches = current_lane_direction.lane->
>         GetOngoingBranches(LaneEnd::kFinish)->size();
> ```

BTW, If I'm not mistaken, wasn't the idea to stop the car after a prediction of its position (assuming it maintains speed) over the time interval determined by the next `DoCalcNextUpdateTime` returns a position that's outside the lane? In this way, the car stops just before the end of the road, not after it has just gone off the end of the road.

---

*[drake/automotive/test/maliput_railcar_test.cc, line 106 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5600#-KfwrQHR5218NeC769pi:-KfwrQHR5218NeC769pj:bjuvhc0) ([raw file](https://github.com/robotlocomotion/drake/blob/a536a916b8c7db975c1b4c777655f475f579f9af/drake/automotive/test/maliput_railcar_test.cc#L106)):*
> ```cpp
>         kStraightRoadLength,                                      /* length */
>         EndpointZ(0, 0, 0, 0));                                   /* z_end  */
>     if (flip_curve_lane) {
> ```

BTW, a simple diagram illustrating the implicit `s`-coordinates for each segment would be helpful here.  Same for the else statement.

---

*[drake/automotive/test/maliput_railcar_test.cc, line 110 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5600#-Kfx1xWTaoFLdqM0Djlt:-Kfx1xWTaoFLdqM0Djlu:b-e1pa1e) ([raw file](https://github.com/robotlocomotion/drake/blob/a536a916b8c7db975c1b4c777655f475f579f9af/drake/automotive/test/maliput_railcar_test.cc#L110)):*
> ```cpp
>           "point.1",                                              /* id     */
>           Endpoint(                                               /* start  */
>               EndpointXy(10 + kCurvedRoadRadius, kCurvedRoadRadius, 1.5 * M_PI),
> ```

BTW, replace with `kStraightRoadLength` since your curved/straight check so heavily relies on it?  Ditto below.

---

*[drake/automotive/test/maliput_railcar_test.cc, line 853 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5600#-KfwzO51PwAQ0fa22aho:-KfwzO51PwAQ0fa22ahp:bjtgsum) ([raw file](https://github.com/robotlocomotion/drake/blob/a536a916b8c7db975c1b4c777655f475f579f9af/drake/automotive/test/maliput_railcar_test.cc#L853)):*
> ```cpp
>   }
> 
>   // Verifies that the car does not stop when it is on the curved lane
> ```

BTW, it seems redundant to do these tests within the interior of both the curved lane and straight road, since `s` is independent of the shape of the road. Conceptually, does this offer any additional coverage?  Would it make more sense to instead test at some other `r`-offset?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5600#-:-Kfx0T7RaUVAgVbifJit:b-efw43w)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Automotive design review is in good shape.

Ditto (as with the other PR) on verbosity, but we can push on that more later.

---

Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/automotive/maliput_railcar.cc, line 374 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5600#-KfwxuVbkIkPJlHdEpAP:-KfxT6hmNAmw0ppVLjzc:b48von4) ([raw file](https://github.com/robotlocomotion/drake/blob/a536a916b8c7db975c1b4c777655f475f579f9af/drake/automotive/maliput_railcar.cc#L374)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

BTW, If I'm not mistaken, wasn't the idea to stop the car after a prediction of its position (assuming it maintains speed) over the time interval determined by the next `DoCalcNextUpdateTime` returns a position that's outside the lane? In this way, the car stops just before the end of the road, not after it has just gone off the end of the road.
</blockquote></details>

Ah, this is regarding the `current_railcar_state->s() > current_lane_direction.lane->length()` below?  I do agree that it should be `>=` and `<=` instead of `>` and `<`.  An epsilon prior to that would be okay as well, but I think not super-important right now.  We're not really depending on precise tolerances right now, so as long as the system doesn't go into an infinite Zeno's loop or chatter around an event, I think we're okay.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5600#-:-KfxTiXuc6YaF0NCUxTw:b-8ta86b)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/automotive/maliput_railcar.cc, line 374 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5600#-KfwxuVbkIkPJlHdEpAP:-Kg-GdMengTfSZZN9hY3:b53vrke) ([raw file](https://github.com/robotlocomotion/drake/blob/a536a916b8c7db975c1b4c777655f475f579f9af/drake/automotive/maliput_railcar.cc#L374)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Ah, this is regarding the `current_railcar_state->s() > current_lane_direction.lane->length()` below?  I do agree that it should be `>=` and `<=` instead of `>` and `<`.  An epsilon prior to that would be okay as well, but I think not super-important right now.  We're not really depending on precise tolerances right now, so as long as the system doesn't go into an infinite Zeno's loop or chatter around an event, I think we're okay.
</blockquote></details>

Great, thanks for the clarity!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5600)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I addressed all reviewer comments. PTAL. Thanks.

---

Review status: 0 of 2 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/automotive/maliput_railcar.cc, line 371 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5600#-KfwytBvGE8oX_1LqI42:-Kg-JkcH0oImWTkTHMGr:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a536a916b8c7db975c1b4c777655f475f579f9af/drake/automotive/maliput_railcar.cc#L371)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

BTW, `at` --> `after`
</blockquote></details>

Done.

---

*[drake/automotive/maliput_railcar.cc, line 374 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5600#-KfwxuVbkIkPJlHdEpAP:-Kg-K5aoUdj6b7DUh5EF:byrjv54) ([raw file](https://github.com/robotlocomotion/drake/blob/a536a916b8c7db975c1b4c777655f475f579f9af/drake/automotive/maliput_railcar.cc#L374)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

Great, thanks for the clarity!
</blockquote></details>

Done. I switched to use `>=` and `<=`.

---

*[drake/automotive/test/maliput_railcar_test.cc, line 106 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5600#-KfwrQHR5218NeC769pi:-Kg-LBZKmrMdZUuur0b-:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a536a916b8c7db975c1b4c777655f475f579f9af/drake/automotive/test/maliput_railcar_test.cc#L106)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

BTW, a simple diagram illustrating the implicit `s`-coordinates for each segment would be helpful here.  Same for the else statement.
</blockquote></details>

Done.

---

*[drake/automotive/test/maliput_railcar_test.cc, line 110 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5600#-Kfx1xWTaoFLdqM0Djlt:-Kg-Lh7FoUAPAp-jvsWx:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a536a916b8c7db975c1b4c777655f475f579f9af/drake/automotive/test/maliput_railcar_test.cc#L110)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

BTW, replace with `kStraightRoadLength` since your curved/straight check so heavily relies on it?  Ditto below.
</blockquote></details>

Done.

---

*[drake/automotive/test/maliput_railcar_test.cc, line 853 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5600#-KfwzO51PwAQ0fa22aho:-Kg-LsSshmM9kb_XcxBR:bxxic6x) ([raw file](https://github.com/robotlocomotion/drake/blob/a536a916b8c7db975c1b4c777655f475f579f9af/drake/automotive/test/maliput_railcar_test.cc#L853)):*
> Conceptually, does this offer any additional coverage?

The straight vs curved lanes are implemented in two different classes: `maliput:monolane::Lane` and `maliput::monolane::ArcLane`. Thus, I believe testing the two cases results in different code paths being executed.

> Would it make more sense to instead test at some other r-offset?

Done. I updated this unit test to evaluate three different r-offsets that span both sides of the s-axis.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5600#-:-Kg-NoKKZ6OuNsxFjXaH:bblnx7v)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5600#-:-Kg0Jn4SV0sKQRZnMd4B:bnfp4nl)*
<!-- Sent from Reviewable.io -->

liangfok commented:
+@ggould-tri for platform review please. Thanks.

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5600#-:-Kg0MxKMHVWjHd83rs3I:b-yqnxar)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5600#-:-Kg0VhGoD8Ik1RMRV0E0:bnfp4nl)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@jwnimmer-tri for final car review please. Thanks.

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5600#-:-Kg0Wxj67iwIUWVMgIYY:b-wjx83k)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
:car:

---

Review status: 1 of 2 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5600#-:-Kg0bK-jKUt_UMZJFvUx:b-8ewamd)*
<!-- Sent from Reviewable.io -->
